### DISCUSSION TITLE
Disable populate for joins I didn't request

### DISCUSSION THREAD

**Discussion by lagnat** [mikro-orm/mikro-orm#3189] [Q&A]
Consider a 1:M query for Author join Books (slightly pseudo-coded)

qb.joinAndSelect('Books')
.where({ Author.name: 'bob' });

The resulting Author:
```
{
    name: 'bob',
    books: [{
        title: 'XXXX',
        ....
        author: {                // I don't want this.
            name: 'bob',         // I only asked for Author join books,
        }                        // not Author join Books join Author
    }]
}
```

**Comment by lagnat** [comment]
@B4nan Sorry to tag you directly but this is more troubling than I initially thought. At issue is that UnitOfWork is populating entities that were not explicitly joined and since I'm using MikroORM for an API, the returned payload is much larger than expected, and in some cases I'm getting a stack overflow on serialization, presumably because UOW added a circular reference.

Is there a way to disable UOW for sub-entities but keep it enabled for top-level entities?

**Reply by B4nan** [reply]
No, this is not possible, and it is a non-goal. If you dont want identity map, dont map to entities and use `qb.execute()` instead. Entities will always work based on identity map, so there will never be two instances for one entity. Also, UoW is not populating anything, it managed entity state and identity, population is a different layer.

If you have an actual problem with serialization or unstable API response, provide a complete reproduction - the serialized form will respect populate hints, so unless you do something weird, there won't be any issues with cycles. But for that to work properly, you should be using the EM, not QB.

> Is there a way to disable UOW for sub-entities but keep it enabled for top-level entities?

There is no such thing as "sub-entity" and "top-level entity", at least not in the eyes of the UoW.

**Comment by lagnat** [comment]
|  If you dont want identity map, dont map to entities and use qb.execute() instead. 

This gives me duplicates and seems not to fully decompose joins into objects.

```
{
    "data": [
        {
            "id": 3,
            "name": "US East (Ohio)",
            "shortName": "US East (Ohio)",
            "originProviderId": 2,
            "code": "us-east-2",
            "backupRegionId": 6,
            "backupOriginRegion": 6,
            "is_active": 1,
            "originInstance": [
                {
                    "id": 5,
                    "name": "steve1",
                    "originAzId": null,
                    "originClassId": 1,
                    "originRegionId": 3,
                    "status": "idle",
                    "instanceUid": null,
                    "statusMessage": "",
                    "ip": null,
                    "isManual": 1,
                    "updatedAt": 1657075550,
                    "owner": null,
                    "amiBuildDate": null,
                    "releaseLabel": "stag",
                    "autoLicensing": 0,
                    "licenseKeyId": 2,
                    "currentOriginClassId": null,
                    "lomRunId": null,
                    "notes": null,
                    "streamLoad": 0,
                    "inService": 1
                },
                {
                    "id": 7,
                    "name": "steve2",
                    "originAzId": null,
                    "originClassId": 1,
                    "originRegionId": 3,
                    "status": "idle",
                    "instanceUid": null,
                    "statusMessage": "",
                    "ip": null,
                    "isManual": 1,
                    "updatedAt": 1657075553,
                    "owner": null,
                    "amiBuildDate": null,
                    "releaseLabel": "stag",
                    "autoLicensing": 0,
                    "licenseKeyId": 4,
                    "currentOriginClassId": null,
                    "lomRunId": null,
                    "notes": null,
                    "streamLoad": 0,
                    "inService": 1
                }
            ]
        },
        {
            "id": 3,
            "name": "US East (Ohio)",
            "shortName": "US East (Ohio)",
            "originProviderId": 2,
            "code": "us-east-2",
            "backupRegionId": 6,
            "backupOriginRegion": 6,
            "is_active": 1,
            "originInstance": [
                {
                    "id": 5,
                    "name": "steve1",
                    "originAzId": null,
                    "originClassId": 1,
                    "originRegionId": 3,
                    "status": "idle",
                    "instanceUid": null,
                    "statusMessage": "",
                    "ip": null,
                    "isManual": 1,
                    "updatedAt": 1657075550,
                    "owner": null,
                    "amiBuildDate": null,
                    "releaseLabel": "stag",
                    "autoLicensing": 0,
                    "licenseKeyId": 2,
                    "currentOriginClassId": null,
                    "lomRunId": null,
                    "notes": null,
                    "streamLoad": 0,
                    "inService": 1
                },
                {
                    "id": 7,
                    "name": "steve2",
                    "originAzId": null,
                    "originClassId": 1,
                    "originRegionId": 3,
                    "status": "idle",
                    "instanceUid": null,
                    "statusMessage": "",
                    "ip": null,
                    "isManual": 1,
                    "updatedAt": 1657075553,
                    "owner": null,
                    "amiBuildDate": null,
                    "releaseLabel": "stag",
                    "autoLicensing": 0,
                    "licenseKeyId": 4,
                    "currentOriginClassId": null,
                    "lomRunId": null,
                    "notes": null,
                    "streamLoad": 0,
                    "inService": 1
                }
            ]
        },
        {
            "id": 4,
            "name": "US East (N. Virginia)",
            "shortName": "US East (N. Virginia)",
            "originProviderId": 2,
            "code": "us-east-1",
            "backupRegionId": 5,
            "backupOriginRegion": 5,
            "is_active": 1,
            "o1__id": null,
            "o1__name": null,
            "o1__origin_az_id": null,
            "o1__origin_class_id": null,
            "o1__origin_region_id": null,
            "o1__status": null,
            "o1__instance_uid": null,
            "o1__status_message": null,
            "o1__ip": null,
            "o1__is_manual": null,
            "o1__updated_at": null,
            "o1__owner": null,
            "o1__ami_build_date": null,
            "o1__release_label": null,
            "o1__auto_licensing": null,
            "o1__license_key_id": null,
            "o1__current_origin_class_id": null,
            "o1__lom_run_id": null,
            "o1__notes": null,
            "o1__stream_load": null,
            "o1__in_service": null,
            "originInstance": []
        },
        {
            "id": 5,
            "name": "US West (N. California)",
            "shortName": "US West (N. California)",
            "originProviderId": 2,
            "code": "us-west-1",
            "backupRegionId": 4,
            "backupOriginRegion": 4,
            "is_active": 1,
            "o1__id": null,
            "o1__name": null,
            "o1__origin_az_id": null,
            "o1__origin_class_id": null,
            "o1__origin_region_id": null,
            "o1__status": null,
            "o1__instance_uid": null,
            "o1__status_message": null,
            "o1__ip": null,
            "o1__is_manual": null,
            "o1__updated_at": null,
            "o1__owner": null,
            "o1__ami_build_date": null,
            "o1__release_label": null,
            "o1__auto_licensing": null,
            "o1__license_key_id": null,
            "o1__current_origin_class_id": null,
            "o1__lom_run_id": null,
            "o1__notes": null,
            "o1__stream_load": null,
            "o1__in_service": null,
            "originInstance": []
        },
        {
            "id": 6,
            "name": "US West (Oregon)",
            "shortName": "US West (Oregon)",
            "originProviderId": 2,
            "code": "us-west-2",
            "backupRegionId": 3,
            "backupOriginRegion": 3,
            "is_active": 1,
            "o1__id": null,
            "o1__name": null,
            "o1__origin_az_id": null,
            "o1__origin_class_id": null,
            "o1__origin_region_id": null,
            "o1__status": null,
            "o1__instance_uid": null,
            "o1__status_message": null,
            "o1__ip": null,
            "o1__is_manual": null,
            "o1__updated_at": null,
            "o1__owner": null,
            "o1__ami_build_date": null,
            "o1__release_label": null,
            "o1__auto_licensing": null,
            "o1__license_key_id": null,
            "o1__current_origin_class_id": null,
            "o1__lom_run_id": null,
            "o1__notes": null,
            "o1__stream_load": null,
            "o1__in_service": null,
            "originInstance": []
        },
        {
            "id": 7,
            "name": "Asia Pacific (Tokyo)",
            "shortName": "Asia Pacific (Tokyo)",
            "originProviderId": 2,
            "code": "ap-northeast-1",
            "backupRegionId": 11,
            "backupOriginRegion": 11,
            "is_active": 1,
            "o1__id": null,
            "o1__name": null,
            "o1__origin_az_id": null,
            "o1__origin_class_id": null,
            "o1__origin_region_id": null,
            "o1__status": null,
            "o1__instance_uid": null,
            "o1__status_message": null,
            "o1__ip": null,
            "o1__is_manual": null,
            "o1__updated_at": null,
            "o1__owner": null,
            "o1__ami_build_date": null,
            "o1__release_label": null,
            "o1__auto_licensing": null,
            "o1__license_key_id": null,
            "o1__current_origin_class_id": null,
            "o1__lom_run_id": null,
            "o1__notes": null,
            "o1__stream_load": null,
            "o1__in_service": null,
            "originInstance": []
        },
        {
            "id": 8,
            "name": "Asia Pacific (Seoul)",
            "shortName": "Asia Pacific (Seoul)",
            "originProviderId": 2,
            "code": "ap-northeast-2",
            "backupRegionId": 12,
            "backupOriginRegion": 12,
            "is_active": 1,
            "o1__id": null,
            "o1__name": null,
            "o1__origin_az_id": null,
            "o1__origin_class_id": null,
            "o1__origin_region_id": null,
            "o1__status": null,
            "o1__instance_uid": null,
            "o1__status_message": null,
            "o1__ip": null,
            "o1__is_manual": null,
            "o1__updated_at": null,
            "o1__owner": null,
            "o1__ami_build_date": null,
            "o1__release_label": null,
            "o1__auto_licensing": null,
            "o1__license_key_id": null,
            "o1__current_origin_class_id": null,
            "o1__lom_run_id": null,
            "o1__notes": null,
            "o1__stream_load": null,
            "o1__in_service": null,
            "originInstance": []
        },
        {
            "id": 9,
            "name": "Asia Pacific (Osaka-Local)",
            "shortName": "Asia Pacific (Osaka-Local)",
            "originProviderId": 2,
            "code": "ap-northeast-3",
            "backupRegionId": null,
            "backupOriginRegion": null,
            "is_active": 0,
            "o1__id": null,
            "o1__name": null,
            "o1__origin_az_id": null,
            "o1__origin_class_id": null,
            "o1__origin_region_id": null,
            "o1__status": null,
            "o1__instance_uid": null,
            "o1__status_message": null,
            "o1__ip": null,
            "o1__is_manual": null,
            "o1__updated_at": null,
            "o1__owner": null,
            "o1__ami_build_date": null,
            "o1__release_label": null,
            "o1__auto_licensing": null,
            "o1__license_key_id": null,
            "o1__current_origin_class_id": null,
            "o1__lom_run_id": null,
            "o1__notes": null,
            "o1__stream_load": null,
            "o1__in_service": null,
            "originInstance": []
        },
        {
            "id": 10,
            "name": "Asia Pacific (Mumbai)",
            "shortName": "Asia Pacific (Mumbai)",
            "originProviderId": 2,
            "code": "ap-south-1",
            "backupRegionId": null,
            "backupOriginRegion": null,
            "is_active": 0,
            "o1__id": null,
            "o1__name": null,
            "o1__origin_az_id": null,
            "o1__origin_class_id": null,
            "o1__origin_region_id": null,
            "o1__status": null,
            "o1__instance_uid": null,
            "o1__status_message": null,
            "o1__ip": null,
            "o1__is_manual": null,
            "o1__updated_at": null,
            "o1__owner": null,
            "o1__ami_build_date": null,
            "o1__release_label": null,
            "o1__auto_licensing": null,
            "o1__license_key_id": null,
            "o1__current_origin_class_id": null,
            "o1__lom_run_id": null,
            "o1__notes": null,
            "o1__stream_load": null,
            "o1__in_service": null,
            "originInstance": []
        },
        {
            "id": 11,
            "name": "Asia Pacific (Singapore)",
            "shortName": "Asia Pacific (Singapore)",
            "originProviderId": 2,
            "code": "ap-southeast-1",
            "backupRegionId": 7,
            "backupOriginRegion": 7,
            "is_active": 1,
            "o1__id": null,
            "o1__name": null,
            "o1__origin_az_id": null,
            "o1__origin_class_id": null,
            "o1__origin_region_id": null,
            "o1__status": null,
            "o1__instance_uid": null,
            "o1__status_message": null,
            "o1__ip": null,
            "o1__is_manual": null,
            "o1__updated_at": null,
            "o1__owner": null,
            "o1__ami_build_date": null,
            "o1__release_label": null,
            "o1__auto_licensing": null,
            "o1__license_key_id": null,
            "o1__current_origin_class_id": null,
            "o1__lom_run_id": null,
            "o1__notes": null,
            "o1__stream_load": null,
            "o1__in_service": null,
            "originInstance": []
        },
        {
            "id": 12,
            "name": "Asia Pacific (Sydney)",
            "shortName": "Asia Pacific (Sydney)",
            "originProviderId": 2,
            "code": "ap-southeast-2",
            "backupRegionId": 8,
            "backupOriginRegion": 8,
            "is_active": 1,
            "o1__id": null,
            "o1__name": null,
            "o1__origin_az_id": null,
            "o1__origin_class_id": null,
            "o1__origin_region_id": null,
            "o1__status": null,
            "o1__instance_uid": null,
            "o1__status_message": null,
            "o1__ip": null,
            "o1__is_manual": null,
            "o1__updated_at": null,
            "o1__owner": null,
            "o1__ami_build_date": null,
            "o1__release_label": null,
            "o1__auto_licensing": null,
            "o1__license_key_id": null,
            "o1__current_origin_class_id": null,
            "o1__lom_run_id": null,
            "o1__notes": null,
            "o1__stream_load": null,
            "o1__in_service": null,
            "originInstance": []
        }
    ]
}```

**Reply by B4nan** [reply]
Yes, this gives you raw data, and when you join a to many relation, you will get a cartesian product.

**Reply by lagnat** [reply]
I was expecting a cartesian product but the first result (and duplicate) had the joined table broken out.  The rest of the results were flat.  Is that expected?

**Reply by lagnat** [reply]
Now that I look closer none of that matters.  I can't use qb.execute() if I want custom types so that's a show stopper.  It is possible to disable UoW?

**Reply by B4nan** [reply]
I really feel like you have XY problem, so instead of answering this, please provide a reproduction of what you are dealing with. I even think you shouldn't be using QB, with EM and populate hints you wouldn't have to deal with any serialization problems.

**Reply by lagnat** [reply]
tl;dr; Thank you for now.  I may take you up on that offer.

I can't thank you enough for being so willing to support this.  My use-case is complicated in that the queries are machine generated based on input from the outside.  QB is pretty much the only way I can do this because I have non-trivial join criteria.  Also note that I am replacing an existing ORM so some of these patterns are not easily changed in my application.  There are many things I love about Mikro and I'm not willing to change my decision to use it :).  It's really down to this one issue where the JSON reply from my service contains unwanted objects.  If the query is A join B, then I would like my JSON response to have an A object with a B child and for that B object to NOT have an A child or any other children for that matter.  This is the "typical" pattern of ORMs that I've used.  

Once again, thanks for being so accessible.  I feel I've over-spent your generosity on this issue so I'll go back and look for another solution.  Disabling UoW probably gets me what I want, so I may experiment with trying to monkey patch it out.

**Reply by B4nan** [reply]
Yeah I understand what you are talking about, and this is exactly what the populate hints are meant to solve. And they require the `populate` option in `em.find` or `em.populate` to be available, but you could as well modify them manually after you obtain the entities from QB. It might be a bit hacky, so I'd rather see a repro so I can give you working example, but you could try to play around with this yourself.

Basically its all about storing serialization context on the internal entity state, it is exactly the `populate` option value - so an array of string (supporting dot notation for nested relations) of what should end up being populated in the response.  

```ts
import { helper } from '@mikro-orm/core';

const entities = await qb.getResult();
entities.forEach(e => {
  helper(e).__serializationContext.populate ??= populate as PopulateOptions<T>[];
});
```

Maybe this could work automatically with QB too, it would be indeed great. I'll try to look into that during weekend.
